Propagate package environment to local workers

fredrikekre opened the pull request:
Propagate package environment to local workers, fixes #28781, closes #42089.

vchuravy commented on stdlib/Distributed/src/managers.jl:
```diff
@@ -462,10 +462,29 @@ function launch(manager::LocalManager, params::Dict, launched::Array, c::Conditi
     bind_to = manager.restrict ? `127.0.0.1` : `$(LPROC.bind_addr)`
+    env = Dict{String,String}(params[:env])
+
+    # JULIA_(LOAD|DEPOT)_PATH are used to populate (LOAD|DEPOT)_PATH on startup,
```
Could we factor this out? We will need to add this to a whole host of other ClusterManager.

My other worry is that this is now inconsistent. Right now the situation is consistently bad, after this code that works well locally might no longer use once you start adding using the SSHManager. Could you add a documentation  section for this behaviour and the SSHManager behaviour?

fredrikekre replied:
> Could we factor this out? We will need to add this to a whole host of other ClusterManager.

Sure, a function that returs the necessary env dict to replicate the package env?

> My other worry is that this is now inconsistent. [...]

Yea I guess. In my opinion this could be updated for `SSHManager` too. The clusters I have experience with share filesystem anyway. `SSHManager` already assumes that the path to the `julia` binary is the same on the remote, and that it is possible to `cd` to the same directory for example, so for remotes that don't share a fs you already now need to do some extra configuration.

fredrikekre replied:
I added docs, and will try something out for SSHManager in a separate PR.

fredrikekre committed:
Distributed: Propagate package environment to local workers.

Local workers now inherit the package environment of the main process,
i.e. the active project, LOAD_PATH, and DEPOT_PATH. This behavior
can be overridden by passing the new `env` keyword argument, or by
passing `--project` in the `exeflags` keyword argument.

Fixes #28781, and closes #42089.

fonsp commented:
About the NEWS.md update, it looks like the ENV variables `JULIA_LOAD_PATH` and `JULIA_DEPOT_PATH` were already propagated in previous Julia versions:

<details>
<summary>Tests on Julia 1.8</summary>

# LOAD_PATH

```
➜  Downloads JULIA_LOAD_PATH="$(pwd)/hello:$JULIA_LOAD_PATH" julia 
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.3 (2022-11-14)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> LOAD_PATH
4-element Vector{String}:
 "/Users/fons/Downloads/hello"
 "@"
 "@v#.#"
 "@stdlib"

julia> import Distributed

julia> Distributed.addprocs(1)
1-element Vector{Int64}:
 2

julia> Distributed.remotecall_eval(Main, 2, :LOAD_PATH)
4-element Vector{String}:
 "/Users/fons/Downloads/hello"
 "@"
 "@v#.#"
 "@stdlib"

```

# DEPOT_PATH

```
➜  Downloads JULIA_DEPOT_PATH="$(pwd)/hello" julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.3 (2022-11-14)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

(@v1.8) pkg> st
  Installing known registries into `~/Downloads/hello`
Status `~/Downloads/hello/environments/v1.8/Project.toml` (empty project)

julia> import Distributed

julia> Distributed.addprocs(1)
1-element Vector{Int64}:
 2

julia> Distributed.remotecall_eval(Main, 2, :DEPOT_PATH)
1-element Vector{String}:
 "/Users/fons/Downloads/hello"
```

</details>

I believe this is because the entire `ENV` is propagated:

```julia
julia> Distributed.remotecall_eval(Main, 2, :ENV) == ENV
true
```

Perhaps the news entry can be reworded? The change is that the *active project from the parent process is propagated*, independently of how it was set (ENV variable, `Pkg.activate`, `--project`, ...). The `DEPOT_PATH` was already being propagated.
